### PR TITLE
Fix reporting traps (faults) to GDB in SE mode

### LINKED ISSUE #123: Reporting traps (faults) to GDB in SE mode is broken
While using gem5 in SE mode I realized that reporting traps to GDB (if connected) is broken / missing at least on 
POWER / RISC-V. 

Consider POWER: https://github.com/gem5/gem5/blob/develop/src/arch/power/faults.cc#L42-L63. `trapToGdb` returns `true` is GDB is connected *and* trap has been reported, otherwise `false`. But the code in `TrapFault::invoke` panics *if* there's a GDB connected and does nothing if there's no GDB at all! I'm pretty sure the logic should be inverse. 

On RISC-V (and possibly other architectures as well) there's no such logic at all, look  at https://github.com/gem5/gem5/blob/develop/src/arch/riscv/faults.cc, so on RISC-V one does not see software breakpoints hitting at all when connecting GDB to debug program running in gem5.

### PR DESCRIPTION
**Pull request by janvrany**
This addresses #123

### CODE REVIEW

**Review by shingarov — COMMENTED** [review]
I believe it *should* be as @janvrany says, and empirically, it happens to be so on every physical RISC-V machine I have around here.

Still, there may be some reason why we are both wrong; @gabemblack what was your rationale for adding lines 167, 168?

**Comment by janvrany** [comment]
Looking at failed test, I do not really understand what went wrong... Has anyone any idea?

**Comment by powerjg** [comment]
I think the VM running the tests died. Can you rebase and force push to update the branch? This will kick off the tests again.

**Comment by shingarov** [comment]
Well I did a CI re-run yesterday and it didn't make any difference.  So yeah, very strange.

**Comment by powerjg** [comment]
Ah, it's hitting a timeout. I'm not totally sure why... I would try running the test that it failed on `timing-cpu_2-cores_no_cache_DualChannelDDR4_2400_arm_boot_test_to-tick-ALL-x86_64-opt`

@mkjost0, do you have any suggestions on diagnosing this?

**Comment by janvrany** [comment]
@powerjg I did run 
```
./build/ALL/gem5.debug -d /tmp/gem5out532w5i9e -re --silent-redirect tests/gem5/arm_boot_tests/configs/arm_boot_exit_run.py --cpu timing --num-cpus 2 --mem-system no_cache --dram-class DualChannelDDR4_2400 --resource-directory tests/gem5/resources --tick-exit 10000000000
```
on my machine and it terminated successfully after ~20 secs.

**Comment by powerjg** [comment]
Let's see if it works this time. Maybe we got unlucky and ran on a really slow CI core or something.

**Comment by janvrany** [comment]
@powerjg I think it got stuck again. Is there a way I can run a single testcase 

timing-cpu_2-cores_no_cache_DualChannelDDR4_2400_arm_boot_test_to-tick-ALL-x86_64-opt

on my machine? (I realized that what I run above was problably the previous one)

**Comment by mkjost0** [comment]
The smallest unit of test we can run is by SuiteUID, which you can run by using `./main.py run --skip-build --uid SuiteUID:tests/gem5/arm_boot_tests/test_linux_boot.py:timing-cpu_2-cores_no_cache_DualChannelDDR4_2400_arm_boot_test_to-tick-ALL-x86_64-opt -vvv` However, it looks like the command you ran above should accomplish the same thing as running this command.

**Comment by janvrany** [comment]
@mkjost0 Thanks. I tried that but still, "works for me". 

I'm not sure what else to try locally, so let me just force-push one commit at time, starting with no-change commit, to see what happens.

**Comment by janvrany** [comment]
OK, so it looks like commit 7e64e7a is causing the hangup. 

In theory that change may cause infinite loop (and therefore CI timeout) for RISC-V SE workloads if that workload depends on (*incorrect*) behavior of PC being increased *before* reporting trap. 

However:
 * regardless of PC value, `invokeSE()` is going to `panic()` so it should not matter and
 * the that that hangs is ARM, so change in RISC-V should not affect it. 

So, I do not understand that's going on here and out of ideas how to debug it.

**Comment by janvrany** [comment]
Ah! It is not ARM test that hangs, it is indeed RISC-V test that hangs - only the CI output is truncated! Now I can debug it. 

Note to myself: this is what fails:
```
./build/ALL/gem5.opt --debug-flags=Exec ./tests/gem5/asmtest/configs/simple_binary_run.py rv64samt-ps-sysclone_d atomic riscv --num-cores 4 --resource-directory ./tests/gem5/resources
```

**Comment by powerjg** [comment]
Maybe we should make this PR be just the power change. I wonder if the gdb issue on RISC-V may be a different problem/solultion.

File: src/arch/riscv/faults.cc [hunk 1]
```diff
@@ -164,8 +164,6 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
         pc_state.set(addr);
         tc->pcState(pc_state);
     } else {
-        inst->advancePC(pc_state);
-        tc->pcState(pc_state);
```
Comment (rogerchang23424) [inline src/arch/riscv/faults.cc:168]: I think move the code section to `SyscallFault::invokeSE(ThreadContext *tc, const StaticInstPtr &inst)` will pass to test CI

**Comment by janvrany** [comment]
@rogerchang23424 Thanks, this is exactly what I did.

**Comment by janvrany** [comment]
OK,  I did what @rogerchang23424 suggested and now all checks pass. I think this is ready for review now.

**Review by rogerchang23424 — APPROVED** [review]

**Review by BobbyRBruce — APPROVED** [review]

**Comment by shingarov** [comment]
Wait, what?

> [janvrany](https://github.com/janvrany) added a commit to janvrany/MachineArithmetic that referenced this pull request 16 hours ago

What does this have to do with https://github.com/janvrany/MachineArithmetic/commit/81b31f18e1a94c26b8a0ccbd8ad7698586eaf788 ??

**Comment by janvrany** [comment]
> 
> What does this have to do with [janvrany/MachineArithmetic@81b31f1](https://github.com/janvrany/MachineArithmetic/commit/81b31f18e1a94c26b8a0ccbd8ad7698586eaf788) ??

Ah, sorry! Indeed this is completely unrelated, I just accidentally pasted wrong URL into commit message of MA. Fixed now.

**Pull request merged**